runtime/runtime-params-estimator/src/replay/fold_db_ops.rs

```diff
@@ -0,0 +1,284 @@
+#[derive(Default)]
+struct State {
+    /// The indent at which this state started
+    indent: usize,
+    /// Keeps track of operations per DB column.
+    ops_cols: BTreeMap<String, BTreeMap<String, usize>>,
+    /// Optionally collect and print detailed statistics for cache hits and misses.
+    cache_stats: Option<CacheStats>,
+}
+
+impl FoldDbOps {
+    pub(super) fn new() -> Self {
+        Self {
+            fold_anchors: HashMap::new(),
+            print_top_level: true,
+            account_filter: None,
+            track_caches: false,
+            states: vec![State::default()],
+            block_hash: None,
+            min_indent: 0,
+            filter_reset_indent: None,
+        }
+    }
+
+    /// Pre-set that folds on chunks.
+    pub(super) fn chunks(self) -> Self {
+        self.fold("apply_transactions", &["receiver", "shard_id"])
+    }
+
+    /// Pre-set that folds on receipts.
+    pub(super) fn receipts(self) -> Self {
+        self.fold("process_receipt", &["receiver", "receipt_id"])
+            .fold("process_transaction", &["tx_hash"])
+            .print_top_level(false)
+    }
+
+    pub(super) fn fold(mut self, anchor: impl Into<String>, printed_fields: &[&str]) -> Self {
+        let fields = printed_fields.into_iter().map(|s| (*s).into()).collect();
+        self.fold_anchors.insert(anchor.into(), fields);
+        self
+    }
+
+    pub(super) fn print_top_level(mut self, yes: bool) -> Self {
+        self.print_top_level = yes;
+        self
+    }
+
+    pub(super) fn with_cache_stats(mut self) -> Self {
+        self.track_caches = true;
+        self.state().cache_stats = Some(CacheStats::default());
+        self
+    }
+
+    pub(super) fn account_filter(mut self, account: Option<String>) -> Self {
+        if account.is_some() {
+            // evaluate nothing if there is a filter, until the filter matches the first time
+            self.min_indent = usize::MAX;
+        }
+        self.account_filter = account;
+        self
+    }
+
+    fn state(&mut self) -> &mut State {
+        self.states.last_mut().expect(EMPTY_STATE_ERR)
+    }
+
+    fn push_state(&mut self, indent: usize) {
+        let cache_stats = if self.track_caches { Some(CacheStats::default()) } else { None };
+        let new_state = State { indent, ops_cols: Default::default(), cache_stats };
+        self.states.push(new_state);
+    }
+
+    fn pop_state(&mut self) -> State {
+        let state = self.states.pop().expect(EMPTY_STATE_ERR);
+        if self.states.is_empty() {
+            self.push_state(0);
+        }
+        state
+    }
+
+    fn skip(&mut self, trace_indent: usize) -> bool {
+        trace_indent < self.min_indent
+    }
+
+    /// Check if indentation has gone back enough to pop current state or reset filter.
+    ///
+    /// Call this before `skip()` to ensure it uses the correct `min_indent`.
+    fn check_indent(&mut self, out: &mut dyn Write, indent: usize) -> anyhow::Result<()> {
```

Comment:
I think the naming could be improved here. `check_indent` doesn't just check indentation, it modifies stack of states, and `skip` doesn't give intuition what is actually skipped.

What do you think about `check_state` (`check_current_state`?) and `skip_eval`?

Reply:
yes my names were really bad here :)
I did the following changes:

`check_indent` -> `update_state`
`skip` -> `skip_eval`